### PR TITLE
Multi index

### PR DESCRIPTION
**Pull request by floridoo**
See #15

### CODE REVIEW

**Comment by floridoo** [comment]
No tests for now. If the approach is ok for you in general, I can add tests and a `StringSliceFieldIndex`.

File: txn.go [hunk 1]
```diff
@@ -147,7 +147,11 @@ func (txn *Txn) Insert(table string, obj interface{}) error {
 
 	// Get the primary ID of the object
 	idSchema := tableSchema.Indexes[id]
-	ok, idVal, err := idSchema.Indexer.FromObject(obj)
+	idIndexer, ok := idSchema.Indexer.(SingleIndexer)
+	if !ok {
```
Comment (armon) [inline txn.go]: I would push this check into the `Validate` of the schema, just so it can be avoided here

File: txn.go [hunk 2]
```diff
@@ -221,7 +269,11 @@ func (txn *Txn) Delete(table string, obj interface{}) error {
...
```
Comment (armon) [inline txn.go]: This check can also move into `Validate`

File: txn.go [hunk 3]
```diff
@@ -167,41 +171,85 @@ func (txn *Txn) Insert(table string, obj interface{}) error {
 
 		// Handle the update by deleting from the index first
 		if update {
-			ok, val, err := indexSchema.Indexer.FromObject(existing)
+			switch indexer := indexSchema.Indexer.(type) {
+			case SingleIndexer:
+				ok, val, err := indexer.FromObject(existing)
```
Comment (armon) [inline txn.go]: May be cleaner to move `ok, vals, err` out of the switch and share the logic between both cases. E.g. the only thing inside the switch is the assign to the variables.

File: txn.go [hunk 4]
```diff
@@ -241,18 +293,37 @@ func (txn *Txn) Delete(table string, obj interface{}) error {
...
+				// Handle non-unique index by computing a unique index.
+				// This is done by appending the primary key which must
+				// be unique anyways.
+				if !indexSchema.Unique {
+					val = append(val, idVal...)
+				}
+				indexTxn.Delete(val)
+			}
+		case MultiIndexer:
+			ok, vals, err := indexer.FromObject(existing)
```
Comment (armon) [inline txn.go]: Same here, just life the variables out of the switch to share the logic, since the multi-value case is a superset

**Comment by armon** [comment]
@floridoo I like the implementation, its moving the right way, left some feedback on cleanups for it!

File: txn.go (hunk 3 above)
Comment (floridoo) [inline txn.go]: @armon updated in next commit. In my opinion the new version is more DRY but less clear, less "idiomatic Go".

**Comment by floridoo** [comment]
@armon Updated and rebased the pull request, added `StringSliceFieldIndex` and tests

**Comment by vieux** [comment]
@floridoo works great for my usecase, thanks!

**Comment by vieux** [comment]
@armon do you plan to merge this anytime soon ? :)

**Comment by aluzzardi** [comment]
Hey @armon, any way to move this forward?

We temporarily switched to @floridoo's fork (docker/swarmkit#1728) to use this feature but we'd very much prefer to move back to mainline.

This feature allows us to handle many to many relationships and without it we had to resort to a very costly linear scan.

This is our use case, maybe there's a better way to do this:  a container can reference multiple networks and a network can be referenced by multiple containers. When removing a network, we want to figure out if any container is referencing it (sanity check to avoid dangling references).

With multi index, we can simply create a container indexer that returns all network IDs the container is referencing. When deleting a network, we just lookup the container table by network ID.

Without it, we have to scan **all** containers (there can be a ton) and manually check if they reference the network.

**Comment by slackpad** [comment]
Hey @floridoo sorry for the delay on this one - it got buried under some things. I'll review this week.

**Comment by slackpad** [comment]
Thanks @floridoo - finally had some time to review and close this out!

**Pull request merged**